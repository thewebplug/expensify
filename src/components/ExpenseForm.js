import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';



const now = moment();
console.log(moment());
class ExpenseForm extends React.Component {
    state = {
        description: this.props.currentExpense ? this.props.currentExpense.description : '',
        note: this.props.currentExpense ? this.props.currentExpense.note : '',
        amount: this.props.currentExpense ? (this.props.currentExpense.amount / 100).toString() : '',
        createdAt:  this.props.currentExpense ? moment(this.props.currentExpense.createdAt) : moment(),
        calenderFocused: false,
        error: undefined
    }

    onDescriptionChange = (e) => {
        this.setState(() => ({description: e.target.value}))
    }

    onNoteChange = (e) => {
        this.setState(() => ({note: e.target.value}))
    }
    onAmountChange = (e) => {
        const amount = e.target.value;
        if(amount.match(/^\d+(\.\d{0,2})?$/)) {
            this.setState(() => ({amount}))
        }
    }

    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({createdAt}));
        }
    }

    onFocusChange = ({focused}) => {
        this.setState(() => ({calenderFocused: focused}));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount){
            const error = 'Oops! Something went wrong'
            this.setState(() =>({error}))
        }
        else{
            this.setState(() => ({error: ''}))
            this.props.onSubmit(({
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            }))
        }
    } 


    render() {
        return(
            <div>
               {this.state.error && <p>{this.state.error}</p>}
               <form onSubmit={this.onSubmit}>
                <input 
                    type="text" 
                    placeholder="Description" 
                    autoFocus  
                    onChange={this.onDescriptionChange}
                    value={this.state.description}
                />
                <input 
                    type="number" 
                    placeholder="Amount" 
                    onChange={this.onAmountChange}
                    value={this.state.amount}
                />
                <SingleDatePicker
                    date={this.state.createdAt}
                    onDateChange={this.onDateChange}
                    focused={this.state.calenderFocused}
                    onFocusChange ={this.onFocusChange}    
                    numberOfMonths={1}
                    isOutsideRange={(day) => false}
                />
                <textarea
                    placeholder="Add a for your expense(optional)" 
                    onChange={this.onNoteChange}
                    value={this.state.note}
                    >
                </textarea>
                <button>Add Expense</button>
               </form>
            </div>
        )
        
    }
}

export default ExpenseForm;