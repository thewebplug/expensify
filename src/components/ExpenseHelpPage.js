import React from 'react';


const ExpenseHelpPage = () => (
    <div>
        Expense Help Page
    </div> 
);

export default ExpenseHelpPage;