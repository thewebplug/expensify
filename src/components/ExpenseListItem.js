import React from 'react';
import { connect } from 'react-redux';
import getVisibleExpenses from '../selectors/expenses';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';


export const ExpenseListItem = (props) => {

    return (
        props.expenses.length === 0 ? 
        <p>No Expenses</p> :
        props.expenses.map((expense) => 

            <div key={expense.id}>
                <Link to={`edit/${expense.id}`}><h3>{expense.description}</h3></Link>
                <p>Amount: {numeral(expense.amount / 100).format('$0,0.00')}</p>
                <p>Created At: {moment(expense.createdAt).format('MMMM Do, YYYY')}</p>
            </div>
    )

    )
}


const mapStateToProps = (state) => {
    return {
        expenses: getVisibleExpenses(state.expenses, state.filters)
    }

}

export default connect(mapStateToProps)(ExpenseListItem);