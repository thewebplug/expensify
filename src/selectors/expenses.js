import moment from "moment";

const getVisibleExpenses = (expenses, {text, sortBy, startDate, endDate}) => {
    return expenses.filter((expense) => {
        const momentCreatedAt = moment(expense.createdAt);

        const startDateMatch = startDate ? momentCreatedAt.isSameOrAfter(startDate, 'day') : true
        const endDateMatch = endDate ? momentCreatedAt.isSameOrBefore(endDate, 'day') : true
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
        
        return startDateMatch && endDateMatch && textMatch;
        
    }).sort((a, b) => {
        if(sortBy === 'date'){
            return a.createdAt < b.createdAt ? 1 : -1;
        }
        else if(sortBy === 'amount') {
            return a.amount > b.amount ? -1 : 1;
        }
    })
}

export default getVisibleExpenses;