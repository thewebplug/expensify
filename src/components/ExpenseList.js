import React from 'react';
import ExpenseListItem from './ExpenseListItem';
const ExpenseList = () => (
    <div>
      <ExpenseListItem />
    </div>
)

export default ExpenseList;
