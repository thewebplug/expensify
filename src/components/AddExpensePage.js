import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';
import { useNavigate } from 'react-router-dom';

export const AddExpensePage = (props) => {
    const onSubmit=(expense) => {
        props.onSubmit(expense);
        props.navigate('/');
    }
    return(
        <div>
            <h1>Add Expenses</h1>
            <ExpenseForm onSubmit={onSubmit}/>
        </div>
    )
};

const mapDispatchToProps = (dispatch) => ({
        onSubmit: (expense) => dispatch(addExpense(expense)),
        navigate: useNavigate()
    })

export default connect(undefined, mapDispatchToProps)(AddExpensePage);