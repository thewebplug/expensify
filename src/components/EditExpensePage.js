import React from 'react';
import { useParams } from 'react-router-dom';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
import { useNavigate } from 'react-router';




export const EditExpensePage = (props) => {
        const {id} = props.useParams;

        const selectedExpense = props.expenses.find((expense) => expense.id === id);

        const onSubmit = (expense) => {
                props.editExpense(selectedExpense.id, expense)
                props.navigate('/')
        }

        const onClick = () => {
                props.removeExpense(selectedExpense.id);
                props.navigate('/')
        }
        return(
                <div>
                        <ExpenseForm
                                currentExpense={selectedExpense} 
                                onSubmit={onSubmit}
                        />
                        <button onClick={onClick}>Remove</button>
                </div>
        )
}
const mapDispatchToProps = (dispatch) => ({
        
        editExpense: (id, expense) => dispatch(editExpense(id, expense)),
        removeExpense: (id) => dispatch(removeExpense({id})),
        navigate: useNavigate(),
        useParams: useParams()
})

const mapStateToProps = (state) => ({
        expenses: state.expenses
})

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);