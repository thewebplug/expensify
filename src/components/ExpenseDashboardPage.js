import React from 'react';
import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';
import ExpenseSummary from './ExpenseSummary';


const ExpenseDashboardPage = () => (
    <div>
        <ExpenseList />
        <ExpenseSummary />
        <ExpenseListFilters />
    </div>
);

export default ExpenseDashboardPage;