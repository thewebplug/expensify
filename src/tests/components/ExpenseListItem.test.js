import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseListItem } from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';


test('should render ExpenseListItem with expenses', () => {
    const wrapper = shallow(<ExpenseListItem expenses={expenses} />)
    expect(wrapper).toMatchSnapshot();
})


test('should render ExpnseListItem with no expenses', () => {
    const wrapper = shallow(<ExpenseListItem expenses={[]}/>)
    expect(wrapper).toMatchSnapshot();
})