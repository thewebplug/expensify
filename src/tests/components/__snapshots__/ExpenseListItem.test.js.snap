// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseListItem with expenses 1`] = `
Array [
  <div
    key="44"
  >
    <Link
      to="edit/44"
    >
      <h3>
        water bill exp
      </h3>
    </Link>
    <p>
      Amount: 
      $10.00
    </p>
    <p>
      Created At: 
      January 11th, 1970
    </p>
  </div>,
  <div
    key="88"
  >
    <Link
      to="edit/88"
    >
      <h3>
        utility bill exp
      </h3>
    </Link>
    <p>
      Amount: 
      $20.00
    </p>
    <p>
      Created At: 
      January 5th, 1970
    </p>
  </div>,
  <div
    key="89"
  >
    <Link
      to="edit/89"
    >
      <h3>
        mend exp
      </h3>
    </Link>
    <p>
      Amount: 
      $14.00
    </p>
    <p>
      Created At: 
      December 27th, 1969
    </p>
  </div>,
]
`;

exports[`should render ExpnseListItem with no expenses 1`] = `
<p>
  No Expenses
</p>
`;
